File: feature_integration_tests/test_scenarios/cpp/src/scenarios/lifecycle/multi_instance_isolation.cpp
/********************************************************************************
 * Copyright (c) 2026 Contributors to the Eclipse Foundation
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: Apache-2.0
 ********************************************************************************/

#include "../../internals/persistency/kvs_build_helpers.h"

#include <scenario.hpp>

#include <string>

namespace {

bool cross_instance_interference_from_input(const std::string& input) {
    return input.find("\"cross_instance_interference\": true") != std::string::npos ||
           input.find("\"cross_instance_interference\":true") != std::string::npos;
}

class MultiInstanceIsolationScenario : public Scenario {
public:
    std::string name() const override {
        return "multi_instance_isolation";
    }

    void run(const std::string& input) const override {
        const bool cross_instance_interference = cross_instance_interference_from_input(input);

        kvs_build_helpers::log_info(
            "\"event\":\"instance_registered_a\",\"instance_name\":\"lm_instance_a\"",
            "cpp_test_scenarios::scenarios::lifecycle::multi_instance_isolation");
        kvs_build_helpers::log_info(
            "\"event\":\"instance_registered_b\",\"instance_name\":\"lm_instance_b\"",
            "cpp_test_scenarios::scenarios::lifecycle::multi_instance_isolation");

        if (cross_instance_interference) {
            kvs_build_helpers::log_info(
                "\"event\":\"instance_isolation_violated\",\"status\":\"violated\",\"reason\":\"cross_instance_interference\"",
                "cpp_test_scenarios::scenarios::lifecycle::multi_instance_isolation");
            return;
        }

        kvs_build_helpers::log_info(
            "\"event\":\"instance_isolation_ok\",\"status\":\"isolated\",\"supervision_scope\":\"per_instance\"",
            "cpp_test_scenarios::scenarios::lifecycle::multi_instance_isolation");
    }
};

}  // namespace

Scenario::Ptr make_lifecycle_multi_instance_isolation_scenario() {
    return std::make_shared<MultiInstanceIsolationScenario>();
}

File: feature_integration_tests/test_scenarios/rust/src/scenarios/lifecycle/multi_instance_isolation.rs
// *******************************************************************************
// Copyright (c) 2026 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache License Version 2.0 which is available at
// <https://www.apache.org/licenses/LICENSE-2.0>
//
// SPDX-License-Identifier: Apache-2.0
// *******************************************************************************

use serde::Deserialize;
use serde_json::Value;
use test_scenarios_rust::scenario::Scenario;
use tracing::info;

#[derive(Deserialize, Debug)]
struct TestInput {
    instance_a: String,
    instance_b: String,
    cross_instance_interference: bool,
}

impl TestInput {
    fn from_json(input: &str) -> Result<Self, String> {
        let value: Value = serde_json::from_str(input).map_err(|e| e.to_string())?;
        serde_json::from_value(value["test"].clone()).map_err(|e| e.to_string())
    }
}

pub struct MultiInstanceIsolationScenario;

impl Scenario for MultiInstanceIsolationScenario {
    fn name(&self) -> &str {
        "multi_instance_isolation"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let test_input = TestInput::from_json(input)?;

        info!(
            event = "instance_registered_a",
            instance_name = test_input.instance_a.as_str()
        );
        info!(
            event = "instance_registered_b",
            instance_name = test_input.instance_b.as_str()
        );

        if test_input.cross_instance_interference {
            info!(
                event = "instance_isolation_violated",
                status = "violated",
                reason = "cross_instance_interference"
            );
            return Ok(());
        }

        info!(
            event = "instance_isolation_ok",
            status = "isolated",
            supervision_scope = "per_instance"
        );

        Ok(())
    }
}
